### PR TITLE
Backward compatible kubernetes_labels behaviour for v3 and v4 roles

### PR DESCRIPTION
**Pull request by smallinsky**
Related: https://github.com/gravitational/teleport/pull/10011

## What 
Align `kubernetes_labels` evaluation behavior for v3 and v4 roles after introducing: https://github.com/gravitational/teleport/pull/9759 

## Why 
Before the _"Fix k8 access - respect kube service labels"_ fix `kubernetes_labels` were never respected. The fix changed that behavior and took into account `kubernetes_labels` during the evaluation of impersonate k8 users/groups. 

#### Broken old role behavior: 

For the following role with Allow and Deny rule where `kubernetes_labels` are not specified in deny rule: 
```
{
  "kind": "role",
  "version": "v3",
  "spec": {
   "allow": {
    "kubernetes_groups": ["system:masters"],
    "kubernetes_users": ["alice@example.com"],
    "kubernetes_labels": { "*": "*" },
   },
   "deny": {
    "kubernetes_groups": ["system:masters"],
    "kubernetes_users": ["alice@example.com"]
   }
  }
 }
 ```
the `deny` rule in `CheckKubeGroupsAndUsers` call is taken into account during the k8 access resource evaluation event if role Deny rules don't contain any `kubernetes_labels`. 

This behavior seems to be inconsistent with current `CheckAccess` logic where if a role has an empty resource label `MatchLabels` is always evaluated to false.  

## Solution:
Make behaviour for current defined v3 and v4 roles backward compatible. Though evaluating MachLabels to true where the service resource label is empty is not consistent with CheckAccess logic. If a new v5 role will be introduce the v5 k8 access logic will automatically become compatible with `CheckAccess` logic.

### CODE REVIEW

**Review by NajiObeid — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

File: lib/services/role.go [hunk 1]
```diff
@@ -1668,10 +1668,28 @@ func NewKubernetesClusterLabelMatcher(clustersLabels map[string]string) RoleMatc
 
 // Match matches a Kubernetes cluster labels against a role.
 func (l *kubernetesClusterLabelMatcher) Match(role types.Role, typ types.RoleConditionType) (bool, error) {
-	ok, _, err := MatchLabels(role.GetKubernetesLabels(typ), l.clusterLabels)
+	ok, _, err := MatchLabels(l.getKubeLabels(role, typ), l.clusterLabels)
 	return ok, trace.Wrap(err)
 }
 
+// getKubeLabels returns kubernetes_labels based on resource version and role type.
+func (l kubernetesClusterLabelMatcher) getKubeLabels(role types.Role, typ types.RoleConditionType) types.Labels {
```
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/services/role.go:1676]: This func will fallback logic is only related to  `kubernetesClusterLabelMatcher` matcher behaviour thus wan't make sure that won't be reused in other use cases.

File: lib/services/role.go [hunk 2]
```diff
@@ -1668,10 +1668,29 @@ func NewKubernetesClusterLabelMatcher(clustersLabels map[string]string) RoleMatc
...
+	// After the introduction of https://github.com/gravitational/teleport/pull/9759 the
+	// kubernetes_labels started to be respected. Former role behavior evaluated deny rules
+	// even if the kubernetes_labels was empty. To preserve this behavior after respecting kubernetes label the label
+	// logic needs to be aligned.
+	// Default wildcard rules should be added to old V3 and V4 roles deny.kubernetes_labels if
+	// deny.kubernetes_labels is empty to ensure that deny rule will be evaluated
+	// even if kubernetes_labels are empty.
+	//
+	// Backward compatibility logic applies only to V3,V4 Role Deny rules.
+	if len(labels) == 0 && typ == types.Deny && (role.GetVersion() == types.V3 || role.GetVersion() == types.V4) {
```
Comment (r0mant) [inline lib/services/role.go:1688]: @smallinsky I'm actually thinking - why do we need to check the role version here? If/when we introduce v5, won't we run into the same problem?

Comment (r0mant) [reply lib/services/role.go:1688]: Cause deny rules always evaluate greedily so I don't think we want users to have to explicitly specify kubernetes_labels in them going forward (that would also be backwards incompatible). Am I missing something?

**Review by r0mant — CHANGES_REQUESTED** [review]
Requesting changes so we don't accidentally merge this until we figure out the question I just left.

**Review by r0mant — APPROVED** [review]

**Pull request merged**